### PR TITLE
[vm] bytecode views now handle generics

### PR DESCRIPTION
**Pull request by sblackshear**
- Moved `derive_type_tag` from the interpreter into the views
- There were some generics-related TODOs in `resolve_in_module` and `normalize_struct` that we can fix using `derive_type_tag`

## Motivation

See above.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

All tests still pass, should be pure refactoring.

### CODE REVIEW

File: language/vm/src/views.rs [hunk 1]
```diff
@@ -645,32 +653,78 @@ impl<'a, T: ModuleAccess> SignatureTokenView<'a, T> {
...
+            TypeParameter(idx) => return type_actual_tags.get(*idx as usize).cloned(),
+            Reference(_) | MutableReference(_) =>
+            // don't derive a tag for references, only structs
+            {
+                return None
+            }
+            Struct(idx, type_actual_tokens) => {
+                if type_actual_tokens.len() != type_actual_tags.len() {
+                    // wrong number of type actuals
+                    return None;
```
Comment (wrwg) [inline language/vm/src/views.rs]: assert instead? This seems to me a programmer error.

Comment (sblackshear) [reply language/vm/src/views.rs]: Assert is more stringent than the old semantics (which returned `None`), but I will try it and see if it breaks anything.

**Comment by vgao1996** [comment]
LGTM. Note that this (or views in general) will need to be redesigned along when we rework the file format.

File: language/vm/src/views.rs (hunk 1 above)
Comment (sblackshear) [reply language/vm/src/views.rs]: Interestingly, a lot of tests fail if you assert instead of returning `None` here! Definitely worth some investigation in the future.

**Comment by sblackshear** [comment]
@vgao1996 really glad to hear that you are planning to do this! I think views could use some redesign to be more ergonomic, but this will be easier if all the relevant logic is in one place.

**Comment by wrwg** [comment]
> @vgao1996 really glad to hear that you are planning to do this! I think views could use some redesign to be more ergonomic, but this will be easier if all the relevant logic is in one place.

IMHO, we need a coherent object model for representing Move programmatically. The views go into the right direction, as they are reflecting Rust's philosophy to distinguish ownership from access logic. But they are still too low level. For example, if I deal with a struct or function programmatically, I shouldn't need to deal with the subtleties of definition vs handle index. Also, if I work with multiple modules, I shouldn't need to think about normalizing types just because `SignatureToken` is not unique across modules.

FWIW, I'm currently working on a higher level object model as part of abstracting the environment for the specification compiler. I should have a PR in a few. I think it is valuable to look at this problem from an application side (the spec compiler) and then compare cards.

**Comment by vgao1996** [comment]
@wrwg That sounds like the right thing to do! @dariorussi & I and have been long questioning the role of the file format and we came to the conclusion that at the end of the day it is a binary data format that should be optimized for serialization & storage. It is not the primary goal for the file format to cater applications that use it at high level. 

In other words, we have quite a few applications that mostly just care about the semantics of the Move module/script carried by the file format and ideally they should convert the file format into their own higher level representation and work on those. (Views are a cheap way to accomplish this.) Unfortunately this is not the case today as we use the file format directly at many places.

Your higher level object model aligns with our vision well and may end up being useful to similar applications.

**Comment by dariorussi** [comment]
@sblackshear I am not sure we need those clean ups as much as we need to review the OM for code. 
Views are intended to give you a referential view over the binary format and that is helpful in some scenario.
Runtime and also verification are more about transformations of the binary format. The binary format carries a load (handles, definitions) that is not useful after loading, it's too low level. And yet some parts of that format are critical at runtime.
As @wrwg, @vgao1996 commented, looking at different API and struct is a better way to think of the problem.
Also, the type tag is a strange abstraction for a type, to me it feels more it's something we compute once and cache away.

**Comment by sblackshear** [comment]
Totally agree with the long-term direction of having a real object model for Move. I think I mischaracterized this diff as a cleanup The primary purpose was more adding some missing functionality to the prover (specifically: getting canonical representations of instantiated type parameters) and the cleanup was a side effect. I will update the summary to reflect this.

> Also, the type tag is a strange abstraction for a type, to me it feels more it's something we compute once and cache away.

I don't understand this part. A `TypeTag` is the canonical representation of a type. Isn't that exactly what you want when comparing representations of types in different modules?

**Comment by dariorussi** [comment]
> > Also, the type tag is a strange abstraction for a type, to me it feels more it's something we compute once and cache away.
> 
> I don't understand this part. A `TypeTag` is the canonical representation of a type. Isn't that exactly what you want when comparing representations of types in different modules?

Maybe it's just a naming issue but overall I'd expect systems to have a concept of Type (once loaded) and use that as the identity for type. The `TypeTag` is a unique representation of a Type for storage purposes (funny enough right now lives in storage) and you can certainly use that as identity (assuming I am understanding your comment). It just seems far fetched to me and a bit of a whack-a-mole story where we grab whatever happens to be unique and use it as the representation of a Type. 
Bottom line we re missing a first class Type in the runtime and I suspect we need a first class Type in every system that tries to work with Move. `TypeTag` is not intended to be that but it can be made to work like that.

**Comment by vgao1996** [comment]
@dariorussi @sblackshear could you remind me what the current state of this is? Do we still want to check this in?
